docs/cri/config.md

```diff
@@ -174,6 +174,11 @@ version = 2
   # limit.
   max_container_log_line_size = 16384
 
+  # log_read_buf_size is the size of the CRI logger read buffer in bytes.
+  # This buffer is used while redirecting container logs into a log file and
+  # has a minimum size of 4096 bytes.
+  log_read_buf_size = 4096
```

Comment:
Are there any other configuration variables that have minimum/maximum values? If so, how do they work when its specified value is out of range?

If possible and consistent with others, I'd like to have a way to tell users that the value is out-of-range, either returning an error or logging the fact.

Reply:
Checked the CRI config variables and I don't see any specific configuration variable that impose a min/max upfront to the client.  Since for readBuffer we are using the bufio package, it does set a min buffer size of [16B](https://cs.opensource.google/go/go/+/refs/tags/go1.18.2:src/bufio/bufio.go;l=41;drc=9956996f6e32a8001457b057bbff3070f869ddbf), 

I am not very sure of how we can impose a max limit here to the clients or if you have a value in mind. In most cases, the expectation is that if clients do not specify the read buffer size or sets it to a small value, we automatically reserve 4KB buffer. but if they specify a very high value, it will cause OOM `fatal error: runtime: out of memory`

In this case, OOM is caused when bufio calls `make()` with given size:

```
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestRedirectLogs$ github.com/containerd/containerd/pkg/cri/io

fatal error: runtime: out of memory

runtime stack:
runtime.throw({0xac25ae?, 0x10000000000?})
	/usr/local/go/src/runtime/panic.go:992 +0x71
runtime.sysMap(0xc000800000, 0x7f1a6abfcd38?, 0x7f1a6abfcda0?)
	/usr/local/go/src/runtime/mem_linux.go:189 +0x11b
runtime.(*mheap).grow(0xfa02a0, 0x8000000?)
	/usr/local/go/src/runtime/mheap.go:1404 +0x225
runtime.(*mheap).allocSpan(0xfa02a0, 0x8000000, 0x0, 0x1)
	/usr/local/go/src/runtime/mheap.go:1170 +0x171
runtime.(*mheap).alloc.func1()
	/usr/local/go/src/runtime/mheap.go:912 +0x65
runtime.systemstack()
	/usr/local/go/src/runtime/asm_amd64.s:469 +0x49

goroutine 6 [running]:
runtime.systemstack_switch()
	/usr/local/go/src/runtime/asm_amd64.s:436 fp=0xc00012d470 sp=0xc00012d468 pc=0x4688a0
runtime.(*mheap).alloc(0x10000000000?, 0x8000000?, 0x0?)
	/usr/local/go/src/runtime/mheap.go:906 +0x65 fp=0xc00012d4b8 sp=0xc00012d470 pc=0x428845
runtime.(*mcache).allocLarge(0xc000180400?, 0x10000000000, 0x1)
	/usr/local/go/src/runtime/mcache.go:213 +0x85 fp=0xc00012d508 sp=0xc00012d4b8 pc=0x418d45
runtime.mallocgc(0x10000000000, 0x9df840, 0x1)
	/usr/local/go/src/runtime/malloc.go:1096 +0x5a5 fp=0xc00012d580 sp=0xc00012d508 pc=0x40f145
runtime.makeslice(0x7f1aa1418108?, 0xab5f0a?, 0xc000180400?)
	/usr/local/go/src/runtime/slice.go:103 +0x52 fp=0xc00012d5a8 sp=0xc00012d580 pc=0x44f472
bufio.NewReaderSize(...)
	/usr/local/go/src/bufio/bufio.go:57
github.com/containerd/containerd/pkg/cri/io.redirectLogs({0xab8d16, 0x9}, {0xb9daf8?, 0xc00041e480}, {0x7f1a780f0f98, 0xc00041e490}, {0xab5f0a, 0x6}, 0xffffffffffffffff, 0x10000000000)
	/home/ec2-user/workspace/containerd/pkg/cri/io/logger.go:138 +0x23d fp=0xc00012d8f0 sp=0xc00012d5a8 pc=0x9914fd
github.com/containerd/containerd/pkg/cri/io.TestRedirectLogs(0xc000137d40)
	/home/ec2-user/workspace/containerd/pkg/cri/io/logger_test.go:273 +0x24b6 fp=0xc00012df70 sp=0xc00012d8f0 pc=0x9946f6
testing.tRunner(0xc000137d40, 0xaff8e8)
	/usr/local/go/src/testing/testing.go:1439 +0x102 fp=0xc00012dfc0 sp=0xc00012df70 pc=0x50ff82
testing.(*T).Run.func1()
	/usr/local/go/src/testing/testing.go:1486 +0x2a fp=0xc00012dfe0 sp=0xc00012dfc0 pc=0x510e2a
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc00012dfe8 sp=0xc00012dfe0 pc=0x46aac1
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1486 +0x35f

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000137ba0, {0xabd4eb?, 0x50f965?}, 0xaff8e8)
	/usr/local/go/src/testing/testing.go:1487 +0x37a
testing.runTests.func1(0xc000137ba0?)
	/usr/local/go/src/testing/testing.go:1839 +0x6e
testing.tRunner(0xc000137ba0, 0xc0002dfcd8)
	/usr/local/go/src/testing/testing.go:1439 +0x102
testing.runTests(0xc0003cff40?, {0xf72e80, 0x2, 0x2}, {0x7f1aa1418108?, 0x40?, 0xf86dc0?})
	/usr/local/go/src/testing/testing.go:1837 +0x457
testing.(*M).Run(0xc0003cff40)
	/usr/local/go/src/testing/testing.go:1719 +0x5d9
main.main()
	_testmain.go:49 +0x1aa
FAIL	github.com/containerd/containerd/pkg/cri/io	0.034s
FAIL
```

Reply:
Sorry, I meant, if someone sets log_read_buf_size=1024 and containerd is ignoring the value since that is < 4096B, it should either log something or exit with non-zero status.